Improving nnet generalization to feature offsets

pegahgh opened the pull request:
This PR is the first implementation for issue #977 
@vimalmanohar  I would be thankful if you go through first round of review for this PR.

danpovey commented:
@pegahgh, do you have any WERs with this?

vimalmanohar commented on src/nnet3bin/nnet3-get-egs.cc:
```diff
@@ -199,6 +210,16 @@ int main(int argc, char *argv[]) {
         }
+        const Matrix<BaseFloat> *cmn_offsets = NULL;
+        if (!utt2cmn_offsets_rspecifier.empty()) {
+          if (offset_reader.HasKey(key)) {
```
!offset_reader

vimalmanohar commented on egs/wsj/s5/steps/nnet3/get_egs.sh:
```diff
@@ -325,10 +325,15 @@ if [ $stage -le 4 ]; then
   # The examples will go round-robin to egs_list.
   $cmd JOB=1:$nj $dir/log/get_egs.JOB.log \
-    nnet3-get-egs --num-pdfs=$num_pdfs $ivector_opt $egs_opts --num-frames=$frames_per_eg "$feats" \
+    nnet3-get-egs --num-pdfs=$num_pdfs $ivector_opt $egs_opts --num-frames=$frames_per_eg "$offsets" "$feats" \
```
I think this should be $offsets without the quotes

vimalmanohar commented on egs/wsj/s5/steps/compute_offsets.sh:
```diff
@@ -0,0 +1,53 @@
+done
+
+$cmd $logdir/compute_rand_offset.log \ 
+  generate-random-cmn-offsets --config=$offsets_config ark:$data/spk2utt scp:$data/feats.scp ark:- | copy-feats --compress=$compress ark:- ark,scp:$offsetdir/offsets.ark,$data/offsets.scp || exit 1;
```
Try to stick to 80 chars per line.

vimalmanohar replied:
You don't seem to have committed this binary.

vimalmanohar commented on src/nnet3/nnet-example.h:
```diff
@@ -44,14 +44,17 @@ struct NnetIo {
   NnetIo(const std::string &name,
          int32 t_begin, const MatrixBase<BaseFloat> &feats);
 
+  NnetIo(const std::string &name,
```
Explain why this is needed and where it is used.

pegahgh commented:
Not yet! I will update you as soon as I get the results.


On Sat, Oct 22, 2016 at 4:17 PM, Daniel Povey <notifications@github.com>
wrote:

> @pegahgh <https://github.com/pegahgh>, do you have any WERs with this?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/1131#issuecomment-255551343>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AI-PtCg2kx22atAZFtfqzlrlV5COnZhyks5q2m9QgaJpZM4Kcqoj>
> .
>

danpovey commented:
@pegahgh, any results on this?

pegahgh commented:
The preliminary results using cmn-offset-scale=0.1 results in 0.2% WER
degradation on SWBD.
Although I used small offset scale, the random offsets are larger than
MFCC. I will try to run new experiment using smaller offset scale.
e.g. MFCC

74.33336 -14.56764  -35.18937 -6.718647

corresponding offsets for these dimensions.

567.3954 -114.6091 -132.4369 -56.41733

The centered covariance of the speaker or pseudo-speaker means is

(1/num_spks) * sum_{i=0, num_spks} (m_i - g)^t (m_i - g)

Large speaker mean, m_i  from global mean g  results in larger covariance,
even if this speaker has few number of frames and it doesn't affect global
CMN.

It may be good idea to use global CMN to generate random offset.

The following figure is the i-vector mapped to 2dim space using PCA for
different offsets of single utterance.

On Sat, Nov 5, 2016 at 4:47 PM, Daniel Povey <notifications@github.com>
wrote:

> @pegahgh <https://github.com/pegahgh>, any results on this?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/1131#issuecomment-258639564>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AI-PtBb6hy4c4ChGh5F4CSvyh94rRorgks5q7OtSgaJpZM4Kcqoj>
> .
>

danpovey commented on src/featbin/generate-random-cmn-offsets.cc:
```diff
@@ -0,0 +1,217 @@
+    // add term gT * g
+    spk_cov.AddVecVec(1.0, global_mean, global_mean);
+    // compute sigma as spk_cov^0.5 for matrix w.r.t its svd decomposition.
+    Matrix<double> sigma(feat_dim, feat_dim);
```
This matrix seems to be initialized as zero.

danpovey commented on src/featbin/generate-random-cmn-offsets.cc:
```diff
@@ -0,0 +1,217 @@
+    spk_cov.AddVecVec(1.0, global_mean, global_mean);
+    // compute sigma as spk_cov^0.5 for matrix w.r.t its svd decomposition.
+    Matrix<double> sigma(feat_dim, feat_dim);
+    sigma.Power(0.5);
```
It's possible that adding an offset generated from the un-scaled sigma will turn out to be 'too strong', and it might be necessary to scale down sigma by some constant  (e.g. 0.5).

danpovey commented on src/featbin/generate-random-cmn-offsets.cc:
```diff
@@ -0,0 +1,217 @@
+    global_mean.Scale(1.0 / global_count);
+
+    // add term -2 * (sum_{i=0}^n m_i^T) g
+    spk_cov.AddVecVec(-2.0, spk_mean_sum, global_mean);
```
At this point you should be dividing by the number of speakers.

pegahgh commented:
I did all of that.
Actually I didn't commit latest changes. I will commit them now.

On Sat, Nov 5, 2016 at 8:10 PM, Daniel Povey <notifications@github.com>
wrote:

> *@danpovey* commented on this pull request.
> ------------------------------
>
> In src/featbin/generate-random-cmn-offsets.cc
> <https://github.com/kaldi-asr/kaldi/pull/1131#pullrequestreview-7324500>:
>
> > +      spk_cov.AddVecVec(1.0, spk_mean, spk_mean);
> +      spk_mean_sum.AddVec(1.0, spk_mean);
> +      global_stats.AddMat(1.0, stats);
> +    }
> +    // global_mean (g) is the global mean for all frames.
> +    Vector<double> global_mean(global_stats.Row(0).Range(0, feat_dim));
> +    double global_count = global_stats(0, feat_dim);
> +    global_mean.Scale(1.0 / global_count);
> +
> +    // add term -2 * (sum_{i=0}^n m_i^T) g
> +    spk_cov.AddVecVec(-2.0, spk_mean_sum, global_mean);
> +
> +    // add term gT * g
> +    spk_cov.AddVecVec(1.0, global_mean, global_mean);
> +    // compute sigma as spk_cov^0.5 for matrix w.r.t its svd decomposition.
> +    Matrix<double> sigma(feat_dim, feat_dim);
>
> This matrix seems to be initialized as zero.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/1131#pullrequestreview-7324500>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AI-PtGSBdnN5BjGtZqeEPA_nf0A24ddoks5q7RrpgaJpZM4Kcqoj>
> .
>

pegahgh commented:
I will use zero-mean while generating random offsets and offset = C^0.5 * x
where x = N(0,1) and scaling sigma by some constant is equivalent to
scaling offsets using cmn-offset-scale.

On Sat, Nov 5, 2016 at 8:17 PM, Daniel Povey <notifications@github.com>
wrote:

> *@danpovey* commented on this pull request.
>
> found a couple bugs
> ------------------------------
>
> In src/featbin/generate-random-cmn-offsets.cc
> <https://github.com/kaldi-asr/kaldi/pull/1131#pullrequestreview-7324560>:
>
> > +      spk_mean_sum.AddVec(1.0, spk_mean);
> +      global_stats.AddMat(1.0, stats);
> +    }
> +    // global_mean (g) is the global mean for all frames.
> +    Vector<double> global_mean(global_stats.Row(0).Range(0, feat_dim));
> +    double global_count = global_stats(0, feat_dim);
> +    global_mean.Scale(1.0 / global_count);
> +
> +    // add term -2 * (sum_{i=0}^n m_i^T) g
> +    spk_cov.AddVecVec(-2.0, spk_mean_sum, global_mean);
> +
> +    // add term gT * g
> +    spk_cov.AddVecVec(1.0, global_mean, global_mean);
> +    // compute sigma as spk_cov^0.5 for matrix w.r.t its svd decomposition.
> +    Matrix<double> sigma(feat_dim, feat_dim);
> +    sigma.Power(0.5);
>
> It's possible that adding an offset generated from the un-scaled sigma
> will turn out to be 'too strong', and it might be necessary to scale down
> sigma by some constant (e.g. 0.5).
> ------------------------------
>
> In src/featbin/generate-random-cmn-offsets.cc
> <https://github.com/kaldi-asr/kaldi/pull/1131#pullrequestreview-7324560>:
>
> > +
> +      // spk_cov is the centered covariance of the speaker means
> +      // as sum_{i=0}^n (m_i - g)^T (m_i - g), where m_i is spk mean for
> +      // speaker i and g is the gloabl mean.
> +      // add sum_{i=0}^n m_i^T * m_i
> +      spk_cov.AddVecVec(1.0, spk_mean, spk_mean);
> +      spk_mean_sum.AddVec(1.0, spk_mean);
> +      global_stats.AddMat(1.0, stats);
> +    }
> +    // global_mean (g) is the global mean for all frames.
> +    Vector<double> global_mean(global_stats.Row(0).Range(0, feat_dim));
> +    double global_count = global_stats(0, feat_dim);
> +    global_mean.Scale(1.0 / global_count);
> +
> +    // add term -2 * (sum_{i=0}^n m_i^T) g
> +    spk_cov.AddVecVec(-2.0, spk_mean_sum, global_mean);
>
> At this point you should be dividing by the number of speakers.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/1131#pullrequestreview-7324560>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AI-PtPoIYI8S14BO6FcLgQKoA4gfK7Mzks5q7RyogaJpZM4Kcqoj>
> .
>

danpovey commented on src/featbin/generate-random-cmn-offsets.cc:
```diff
@@ -0,0 +1,222 @@
+    double global_count = global_stats(0, feat_dim);
+    global_mean.Scale(1.0 / global_count);
+
+    Vector<double> global_mean2(spk_mean_sum); 
```
not a good choice of variable name.

danpovey commented on src/featbin/generate-random-cmn-offsets.cc:
```diff
@@ -0,0 +1,222 @@
+    spk_cov.AddVecVec(-2.0, spk_mean_sum, global_mean2);
+    
+    // add term gT * g
+    spk_cov.AddVecVec(1.0, global_mean2, global_mean2);
```
this 1.0 should be num_spks. 
This may be the cause of the high variance you observed.

danpovey replied:
... also, I think it would be best if you add some 'self-testing' code to this program.  E.g. take the covariance of the speakers after adding the CMN offsets, multiply it on each side by the inverse-stddev of the original speaker covariance, print out the eigenvalues, and make sure they are close to 1 (if it's been done in such a way that the speaker covariance is supposed to be the same).

pegahgh commented:
Thanks Dan.
I fixed the bugs and added self-testing code to this program.
The figure shows eigenvalues for C^-1 * offset_C, which seems to the
eigenvalues are close to 1. Each offsets for each speaker counted as
separate speaker in computing offset speaker covariance.



On Sun, Nov 6, 2016 at 1:16 PM, Daniel Povey <notifications@github.com>
wrote:

> *@danpovey* commented on this pull request.
> ------------------------------
>
> In src/featbin/generate-random-cmn-offsets.cc
> <https://github.com/kaldi-asr/kaldi/pull/1131>:
>
> > +      spk_mean_sum.AddVec(1.0, spk_mean);
> +      global_stats.AddMat(1.0, stats);
> +    }
> +    // global_mean (g) is the global mean for all frames.
> +    Vector<double> global_mean(global_stats.Row(0).Range(0, feat_dim));
> +
> +    double global_count = global_stats(0, feat_dim);
> +    global_mean.Scale(1.0 / global_count);
> +
> +    Vector<double> global_mean2(spk_mean_sum);
> +    global_mean2.Scale(1.0 / num_spks);
> +    // add term -2 * (sum_{i=0}^n m_i^T) g
> +    spk_cov.AddVecVec(-2.0, spk_mean_sum, global_mean2);
> +
> +    // add term gT * g
> +    spk_cov.AddVecVec(1.0, global_mean2, global_mean2);
>
> ... also, I think it would be best if you add some 'self-testing' code to
> this program. E.g. take the covariance of the speakers after adding the CMN
> offsets, multiply it on each side by the inverse-stddev of the original
> speaker covariance, print out the eigenvalues, and make sure they are close
> to 1 (if it's been done in such a way that the speaker covariance is
> supposed to be the same).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/1131>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AI-PtAr5FknaEe69JEzo4ARV4vfs1Uywks5q7hl_gaJpZM4Kcqoj>
> .
>

pegahgh commented on egs/wsj/s5/steps/nnet3/train_dnn.py:
```diff
@@ -298,13 +299,17 @@ def TrainNewModels(dir, iter, srand, num_jobs, num_archives_processed, num_archi
         frame = (k / num_archives) % frames_per_eg
+        offset_num = -1
+        if num_cmn_offsets > 0:
+          offset_num = (archive_index - 1) % num_cmn_offsets
```
Dan, I fixed this issue in my local repo as follows, but I didn't commit the change.
epoch_index = num_archives_processed / (num_archives * frames_per_eg)
offset_num = epoch_index % num_cmn_offsets

e.g. num_archives_to_process = num_epochs * frames_per_eg * num_archives =>
epoch_index = num_archives_processed / ( frames_per_eg * num_archives)

Isn't that what we expect?
Also I think we need to process the data and its perturbed versions randomly.
e.g. in ubm offset model, while adding offset in 2nd epoch, I got large log-likelihood degradation in the beginning of 2nd epoch.
Any idea?

danpovey commented:
That formula should be better was was checked in, but my formula might be
even better as it mixes together different offsets at the same time.  In
your case you train on non-offsetted data for the 1st epoch, and then
suddenly start training on the offsetted data.
But if the log-like degradation is very large (e.g. more than 0.1 or so),
it may indicate that too much variance is being added to the features, or
maybe it indicates some other problem.

On Fri, Nov 25, 2016 at 11:25 PM, pegahgh <notifications@github.com> wrote:

> *@pegahgh* commented on this pull request.
> ------------------------------
>
> In egs/wsj/s5/steps/nnet3/train_dnn.py
> <https://github.com/kaldi-asr/kaldi/pull/1131>:
>
> > @@ -298,13 +299,17 @@ def TrainNewModels(dir, iter, srand, num_jobs, num_archives_processed, num_archi
>                                                 # the other indexes from.
>          archive_index = (k % num_archives) + 1 # work out the 1-based archive index.
>          frame = (k / num_archives) % frames_per_eg
> +        offset_num = -1
> +        if num_cmn_offsets > 0:
> +          offset_num = (archive_index - 1) % num_cmn_offsets
>
> Dan, I fixed this issue in my local repo as follows, but I didn't commit
> the change.
> epoch_index = num_archives_processed / (num_archives * frames_per_eg)
> offset_num = epoch_index % num_cmn_offsets
>
> e.g. num_archives_to_process = num_epochs * frames_per_eg * num_archives =>
> epoch_index = num_archives_processed / ( frames_per_eg * num_archives)
>
> Isn't that what we expect?
> Also I think we need to process the data and its perturbed versions
> randomly.
> e.g. in ubm offset model, while adding offset in 2nd epoch, I got large
> log-likelihood degradation in the beginning of 2nd epoch.
> Any idea?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/1131>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADJVuwP9E0VeIqZbV1vrsQ1ioQ4LisX7ks5rB7S7gaJpZM4Kcqoj>
> .
>

pegahgh commented:
Dan, I tested generalized shift on TDNN+xent and frames_per_offset is 8,
where we have 3 shifts of data in Chain model. I can start testing that on
chain model.
Also I think we should use k instead on num_archives_processed in the
formula to be compatible with frame number, am I right?

The range for num_archives_processed / num_archives is [0, frames_per_eg *
num_epochs] and and based on proposed formula, the offset for fixed
archive_index = f, and fixed shift of data 's' in epoch epoch_index,
(epoch_index * frames_per_eg + s) % num_offsets = epoch_index *
(frames_per_eg % num_offsets + (s % num_offsets). To have different offset
for same shift in different epochs, frames_per_eg % num_offsets != 0, (if
num_offsets and frames_per_eg would be coprime is one of the cases)!!

For TDNN+xent, frames_per_eg=8, num_offsets=3 and num_epochs=3 should be
appropriate!

Do you agree?

On Sat, Nov 26, 2016 at 2:28 AM, Daniel Povey <notifications@github.com>
wrote:

> *@danpovey* commented on this pull request.
> ------------------------------
>
> In egs/wsj/s5/steps/nnet3/train_dnn.py
> <https://github.com/kaldi-asr/kaldi/pull/1131>:
>
> > @@ -298,13 +299,17 @@ def TrainNewModels(dir, iter, srand, num_jobs, num_archives_processed, num_archi
>                                                 # the other indexes from.
>          archive_index = (k % num_archives) + 1 # work out the 1-based archive index.
>          frame = (k / num_archives) % frames_per_eg
> +        offset_num = -1
> +        if num_cmn_offsets > 0:
> +          offset_num = (archive_index - 1) % num_cmn_offsets
>
> Actually, I think it might be better to do:
>
> offset_num = ((num_archives_processed / num_archives) + (num_archives_processed % num_archives)) % num_cmn_offsets
>
> and ensure that num_cmn_offsets is always coprime to 3 (i.e. not a
> multiple of 3).
> Your formula was only going to a new offset index after you went through
> all 3 shifts of the data-- it would be better to use a new offset each time
> you shift the same archive differently, to enhance how different the
> different versions of the data are.
> Ensuring that the num_cmn_offsets is coprime to the number of shifts (3)
> ensures that this formula doesn't give you only some combinations of
> cmn_offsets and shifts.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/1131>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AI-PtA0FNb6KDwuqVNJXudTjpks87HPSks5rB9-GgaJpZM4Kcqoj>
> .
>

danpovey commented:
You seem to be pointing out that my formula is problematic if the
num_offsets is not also coprime to frames_per_eg.  That seems valid.  (So
for regular xent TDNN training it would have to coprime to frames_per_eg,
and for chain training it would have to be coprime to num_shifts(=3)).  Or
use a different formula.  I did not understand the specific details of what
you were saying.  But anyway, try to come up with a formula that would
satisfy all the requirements, or ensure that the num-offsets is coprime to
whatever it needs to be coprime to.  This idea may not even work.

On Sat, Nov 26, 2016 at 3:37 PM, pegahgh <notifications@github.com> wrote:

> Dan, I tested generalized shift on TDNN+xent and frames_per_offset is 8,
> where we have 3 shifts of data in Chain model. I can start testing that on
> chain model.
> Also I think we should use k instead on num_archives_processed in the
> formula to be compatible with frame number, am I right?
>
> The range for num_archives_processed / num_archives is [0, frames_per_eg *
> num_epochs] and and based on proposed formula, the offset for fixed
> archive_index = f, and fixed shift of data 's' in epoch epoch_index,
> (epoch_index * frames_per_eg + s) % num_offsets = epoch_index *
> (frames_per_eg % num_offsets + (s % num_offsets). To have different offset
> for same shift in different epochs, frames_per_eg % num_offsets != 0, (if
> num_offsets and frames_per_eg would be coprime is one of the cases)!!
>
> For TDNN+xent, frames_per_eg=8, num_offsets=3 and num_epochs=3 should be
> appropriate!
>
> Do you agree?
>
> On Sat, Nov 26, 2016 at 2:28 AM, Daniel Povey <notifications@github.com>
> wrote:
>
> > *@danpovey* commented on this pull request.
> > ------------------------------
> >
> > In egs/wsj/s5/steps/nnet3/train_dnn.py
> > <https://github.com/kaldi-asr/kaldi/pull/1131>:
> >
> > > @@ -298,13 +299,17 @@ def TrainNewModels(dir, iter, srand, num_jobs,
> num_archives_processed, num_archi
> > # the other indexes from.
> > archive_index = (k % num_archives) + 1 # work out the 1-based archive
> index.
> > frame = (k / num_archives) % frames_per_eg
> > + offset_num = -1
> > + if num_cmn_offsets > 0:
> > + offset_num = (archive_index - 1) % num_cmn_offsets
> >
> > Actually, I think it might be better to do:
> >
> > offset_num = ((num_archives_processed / num_archives) +
> (num_archives_processed % num_archives)) % num_cmn_offsets
> >
> > and ensure that num_cmn_offsets is always coprime to 3 (i.e. not a
> > multiple of 3).
> > Your formula was only going to a new offset index after you went through
> > all 3 shifts of the data-- it would be better to use a new offset each
> time
> > you shift the same archive differently, to enhance how different the
> > different versions of the data are.
> > Ensuring that the num_cmn_offsets is coprime to the number of shifts (3)
> > ensures that this formula doesn't give you only some combinations of
> > cmn_offsets and shifts.
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/kaldi-asr/kaldi/pull/1131>, or mute the thread
> > <https://github.com/notifications/unsubscribe-auth/AI-
> PtA0FNb6KDwuqVNJXudTjpks87HPSks5rB9-GgaJpZM4Kcqoj>
> > .
> >
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/1131#issuecomment-263084311>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ADJVu0yAT0Copcd3TcQwD2ieni9EqKy0ks5rCJh0gaJpZM4Kcqoj>
> .
>

danpovey commented:
@pegahgh, is this PR a duplicate and should it be deleted?

pegahgh commented:
Yes, the new PR is #1323.

On Mon, Jan 23, 2017 at 1:48 AM, Daniel Povey <notifications@github.com>
wrote:

> @pegahgh <https://github.com/pegahgh>, is this PR a duplicate and should
> it be deleted?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/1131#issuecomment-274412552>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AI-PtDHisWlNPVLLx1IOkv-FuwHtnwT3ks5rVE1bgaJpZM4Kcqoj>
> .
>
